controller/apps.go

```diff
@@ -0,0 +1,464 @@
+}
+
+func (c *AppsController) getSpaceNameFromSpaceID(ctx context.Context, spaceID uuid.UUID) (*string, error) {
+	// TODO - add a cache in AppsController - but will break if user can change space name
+	// use WIT API to convert Space UUID to Space name
+	osioclient := getAndCheckOSIOClient(ctx)
+
+	osioSpace, err := osioclient.GetSpaceByID(ctx, spaceID)
+	if err != nil {
+		return nil, err
+	}
+	return osioSpace.Attributes.Name, nil
+}
+
+func getNamespaceName(ctx context.Context) (*string, error) {
+
+	osioclient := getAndCheckOSIOClient(ctx)
+	kubeSpaceAttr, err := osioclient.GetNamespaceByType(ctx, nil, "user")
+	if err != nil {
+		return nil, err
+	}
+	if kubeSpaceAttr == nil || kubeSpaceAttr.Name == nil {
+		return nil, witerrors.NewNotFoundError("namespace", "user")
+	}
+
+	return kubeSpaceAttr.Name, nil
+}
+
+func getUser(authClient authservice.Client, ctx context.Context) (*authservice.User, error) {
+	// get the user definition (for cluster URL)
+	resp, err := authClient.ShowUser(ctx, authservice.ShowUserPath(), nil, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	defer resp.Body.Close()
+
+	respBody, err := ioutil.ReadAll(resp.Body)
+
+	status := resp.StatusCode
+	if status < 200 || status > 300 {
+		return nil, fmt.Errorf("Failed to GET user due to status code %d", status)
+	}
+
+	var respType authservice.User
+	err = json.Unmarshal(respBody, &respType)
+	if err != nil {
+		return nil, err
+	}
+	return &respType, nil
+}
+
+func getTokenData(authClient authservice.Client, ctx context.Context, forService string) (*authservice.TokenData, error) {
+
+	resp, err := authClient.RetrieveToken(ctx, authservice.RetrieveTokenPath(), forService, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	defer resp.Body.Close()
+
+	respBody, err := ioutil.ReadAll(resp.Body)
+
+	status := resp.StatusCode
+	if status < 200 || status > 300 {
+		return nil, errors.New("Failed to GET user due to status code " + string(status))
+	}
+
+	var respType authservice.TokenData
+	err = json.Unmarshal(respBody, &respType)
+	if err != nil {
+		return nil, err
+	}
+	return &respType, nil
+}
+
+// getKubeClient createa kube client for the appropriate cluster assigned to the current user.
+// many different errors are possible, so controllers should call getAndCheckKubeClient() instead
+func (c *AppsController) getKubeClient(ctx context.Context) (kubernetes.KubeClientInterface, error) {
+
+	// create Auth API client
+	authClient, err := auth.CreateClient(ctx, c.Config)
+	if err != nil {
+		log.Error(ctx, nil, "error accessing Auth server"+tostring(err))
+		return nil, err
+	}
+
+	authUser, err := getUser(*authClient, ctx)
+	if err != nil {
+		log.Error(ctx, nil, "error accessing Auth server"+tostring(err))
+		return nil, err
+	}
+
+	if authUser == nil || authUser.Data.Attributes.Cluster == nil {
+		log.Error(ctx, nil, "error getting user from Auth server:"+tostring(authUser))
+		return nil, fmt.Errorf("error getting user from Auth Server: %s", tostring(authUser))
```

Comment:
Please don't use `fmt.Errorf` but `errs.Errorf`.

Reply:
done (in followup PR)

Reply:
done (in followup PR)